File: Assets/ProjectGame/Lib/data/enum/Faction/FactionID.py
from enum import IntEnum
from typing import Callable, List, Any


class FactionID(IntEnum):
    NONE = 0  # デフォルト値（C#のNoneに相当）
    Human = 1  # 人類
    Werewolf = 2  # 人狼
    MAX = 3

def to_int(id: "FactionID") -> int:
    return int(id)


def to_FactionID(id: int) -> "FactionID":
    return FactionID(id)


def to_index(id: "FactionID") -> int:
    return int(id) - 1


def for_id(action: Callable[["FactionID"], None]):
    if action is None:
        raise ValueError("action cannot be None")
    start = FactionID.Human.value
    for i in range(start, FactionID.MAX.value):
        try:
            value = FactionID(i)
            action(value)
        except ValueError:
            continue  # 未定義の値はスキップ


def find_all(predicate: Callable[["FactionID"], bool]) -> List["FactionID"]:
    if predicate is None:
        raise ValueError("predicate cannot be None")
    results: List["FactionID"] = []
    start = FactionID.Human.value
    for i in range(start, FactionID.MAX.value):
        try:
            value = FactionID(i)
            if predicate(value):
                results.append(value)
        except ValueError:
            continue
    return results


def find(predicate: Callable[["FactionID"], bool]) -> "FactionID":
    if predicate is None:
        raise ValueError("predicate cannot be None")
    start = FactionID.Human.value
    for i in range(start, FactionID.MAX.value):
        try:
            value = FactionID(i)
            if predicate(value):
                return value
        except ValueError:
            continue
    return FactionID.NONE

File: Assets/ProjectGame/Lib/data/enum/Faction/FactionID.js
// FactionID - Pure JavaScript Version
export const FactionID = {
    None: 0,  // デフォルト値（C#互換）
    Human: 1,  // 人類
    Werewolf: 2,  // 人狼
    Max: 3
};

export const FactionIDExtensions = {
    /**
     * Enumを数値に変換
     */
    toInt(id) {
        return Number(id);
    },

    /**
     * 数値をEnumに変換
     */
    toFactionID(id) {
        return id;
    },

    /**
     * 0-based indexに変換
     */
    toIndex(id) {
        return Number(id) - 1;
    },

    /**
     * すべてのIDに対して処理を実行
     */
    forID(action) {
        if (typeof action !== 'function') {
            throw new Error('action must be a function');
        }
        const start = FactionID.Human;
        const max = FactionID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(FactionID).includes(id)) {
                action(id);
            }
        }
    },

    /**
     * 条件に合うすべてのIDを返す
     */
    findAll(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const results = [];
        const start = FactionID.Human;
        const max = FactionID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(FactionID).includes(id)) {
                if (predicate(id)) {
                    results.push(id);
                }
            }
        }
        return results;
    },

    /**
     * 条件に合う最初のIDを返す（見つからなければ None）
     */
    find(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const start = FactionID.Human;
        const max = FactionID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(FactionID).includes(id)) {
                if (predicate(id)) {
                    return id;
                }
            }
        }
        return FactionID.None;
    }
};
